xds/src/main/java/io/grpc/xds/EnvoyServerProtoData.java

```diff
@@ -0,0 +1,303 @@
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Defines gRPC data types for Envoy protobuf messages used in xDS protocol on the server side,
+ * similar to how {@link EnvoyProtoData} defines it for the client side.
+ */
+final class EnvoyServerProtoData {
+
+  // Prevent instantiation.
+  private EnvoyServerProtoData() {
+  }
+
+  static final class CidrRange {
+    private final String addressPrefix;
+    private final int prefixLen;
+
+    @VisibleForTesting
+    CidrRange(String addressPrefix, int prefixLen) {
+      this.addressPrefix = addressPrefix;
+      this.prefixLen = prefixLen;
+    }
+
+    static CidrRange fromEnvoyProtoCidrRange(
+        io.envoyproxy.envoy.api.v2.core.CidrRange proto) {
+      return new CidrRange(proto.getAddressPrefix(), proto.getPrefixLen().getValue());
+    }
+
+    public String getAddressPrefix() {
+      return addressPrefix;
+    }
+
+    public int getPrefixLen() {
+      return prefixLen;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      CidrRange cidrRange = (CidrRange) o;
+      return prefixLen == cidrRange.prefixLen
+          && java.util.Objects.equals(addressPrefix, cidrRange.addressPrefix);
+    }
+
+    @Override
+    public int hashCode() {
+      return java.util.Objects.hash(addressPrefix, prefixLen);
+    }
+
+    @Override
+    public String toString() {
+      return "CidrRange{"
+          + "addressPrefix='" + addressPrefix + '\''
+          + ", prefixLen=" + prefixLen
+          + '}';
+    }
+  }
+
+  /**
+   * Corresponds to Envoy proto message
+   * {@link io.envoyproxy.envoy.api.v2.listener.FilterChainMatch}.
+   */
+  static final class FilterChainMatch {
+    private final int destinationPort;
+    private final List<CidrRange> prefixRanges;
+    private final List<String> applicationProtocols;
+
+    private FilterChainMatch(int destinationPort,
+        List<CidrRange> prefixRanges, List<String> applicationProtocols) {
+      this.destinationPort = destinationPort;
+      this.prefixRanges = prefixRanges;
+      this.applicationProtocols = applicationProtocols;
+    }
+
+    static FilterChainMatch fromEnvoyProtoFilterChainMatch(
+        io.envoyproxy.envoy.api.v2.listener.FilterChainMatch proto) {
+      List<CidrRange> prefixRanges = new ArrayList<>();
+      for (io.envoyproxy.envoy.api.v2.core.CidrRange range : proto.getPrefixRangesList()) {
+        prefixRanges.add(CidrRange.fromEnvoyProtoCidrRange(range));
+      }
+      List<String> applicationProtocols = new ArrayList<>();
+      for (String appProtocol  : proto.getApplicationProtocolsList()) {
+        applicationProtocols.add(appProtocol);
+      }
+      return new FilterChainMatch(
+          proto.getDestinationPort().getValue(),
+          Collections.unmodifiableList(prefixRanges),
```

Comment:
i expected to have it (unmodifiableList) in the constructor, just in case we have another factory method it can be missed.